internal/nostr/nostr.go

```diff
@@ -614,53 +662,112 @@ func (svc *Service) startSubscription(ctx context.Context, subscription *Subscri
+				"relayUrl":       subscription.RelayUrl,
+			}).Error("Subscription stopped due to relay error, reconnecting in 5s...")
+			time.Sleep(5 * time.Second) // sleep for 5 seconds
+			continue
+		} else {
+			if (subscription.RequestEvent != nil) {
+				if (subscription.RequestEventDB.State == "") {
+					subscription.RequestEventDB.State = REQUEST_EVENT_PUBLISH_FAILED
+				}
+				svc.db.Save(&subscription.RequestEventDB)
 			}
+			svc.Logger.WithFields(logrus.Fields{
+				"subscriptionId": subscription.ID,
+				"relayUrl":       subscription.RelayUrl,
+			}).Info("Stopping subscription")
+			break
 		}
 	}
 }
 
-func (svc *Service) processEvents(ctx context.Context, subscription *Subscription, sub *nostr.Subscription) error {
+func (svc *Service) publishEvent(ctx context.Context, subscription *Subscription, sub *nostr.Subscription) {
+	err := sub.Relay.Publish(ctx, *subscription.RequestEvent)
+	if err != nil {
+		// TODO: notify user about publish failure
+		svc.Logger.WithError(err).WithFields(logrus.Fields{
+			"subscriptionId": subscription.ID,
+			"relayUrl":       subscription.RelayUrl,
+		}).Error("Failed to publish to relay")
+		sub.Unsub()
+	} else {
+		svc.Logger.WithFields(logrus.Fields{
+			"status":  REQUEST_EVENT_PUBLISH_CONFIRMED,
+			"eventId": subscription.RequestEvent.ID,
+		}).Info("Published request event successfully")
+		subscription.RequestEventDB.State = REQUEST_EVENT_PUBLISH_CONFIRMED
+	}
+}
+
+func (svc *Service) persistResponseEvent(event *nostr.Event, subscription *Subscription, sub *nostr.Subscription) {
+	svc.Logger.WithFields(logrus.Fields{
+		"eventId":        event.ID,
+		"eventKind":      event.Kind,
+		"requestEventId": subscription.RequestEvent.ID,
+	}).Info("Received response event")
+	responseEvent := ResponseEvent{
+		NostrId:   event.ID,
+		Content:   event.Content,
+		RepliedAt: event.CreatedAt.Time(),
+		RequestId: &subscription.RequestEventDB.ID,
+	}
+	svc.db.Save(&responseEvent)
+	if subscription.WebhookUrl != "" {
+		svc.postEventToWebhook(event, subscription.WebhookUrl)
+	} else {
+		subscription.EventChan <- event
+	}
+	sub.Unsub()
+}
+
+func (svc *Service) persistSubscribedEvent(event *nostr.Event, subscription *Subscription, sub *nostr.Subscription) {
+	svc.Logger.WithFields(logrus.Fields{
+		"eventId":        event.ID,
+		"eventKind":      event.Kind,
+		"subscriptionId": subscription.ID,
+	}).Info("Received event")
+	responseEvent := ResponseEvent{
+		NostrId:        event.ID,
+		Content:        event.Content,
+		RepliedAt:      event.CreatedAt.Time(),
+		SubscriptionId: &subscription.ID,
+	}
+	svc.db.Save(&responseEvent)
+	svc.postEventToWebhook(event, subscription.WebhookUrl)
+}
+
+func (svc *Service) processEvents(ctx context.Context, subscription *Subscription, sub *nostr.Subscription, onReceiveEOS func(ctx context.Context, subscription *Subscription, sub *nostr.Subscription), persistEvent func(event *nostr.Event, subscription *Subscription, sub *nostr.Subscription)) error {
```

Comment:
if we are ok to go this way, I would at least extract the functions as types 

e.g. type OnReceiveEOSFunc = func(ctx context.Context, subscription *Subscription, sub *nostr.Subscription)

Also, two args like `sub` and `subscription` is not very good - consider someone reading the code for the first time and trying to understand what the difference is

Reply:
> e.g. type OnReceiveEOSFunc = func(ctx context.Context, subscription *Subscription, sub *nostr.Subscription)

Done!

> Also, two args like sub and subscription is not very good

Yup, was thinking about that, changed to only pass `subscription`